src/core/exec-invoke.c

```diff
@@ -4930,23 +4977,30 @@ int exec_invoke(
         if (params->cgroup_path) {
                 _cleanup_free_ char *p = NULL;
 
-                r = exec_params_get_cgroup_path(params, cgroup_context, &p);
+                r = exec_params_get_cgroup_path(params, cgroup_context, params->cgroup_path, &p);
                 if (r < 0) {
                         *exit_status = EXIT_CGROUP;
                         return log_exec_error_errno(context, params, r, "Failed to acquire cgroup path: %m");
                 }
 
-                r = cg_attach_everywhere(params->cgroup_supported, p, 0);
+                /* We cannot spawn the main service process into the subcgroup as it needs to unshare the cgroup
+                 * namespace first if one is configured to make sure the root of the cgroup namespace is the service
+                 * cgroup and not the subcgroup. However, when running control commands on a live service, the
+                 * commands have to be spawned inside a subcgroup, otherwise we violate the no inner processes rule
+                 * of cgroupv2. */
+                const char *cgtarget = exec_params_needs_control_subcgroup(params) ? p : params->cgroup_path;
```

Comment:
How silly would it be to setns(2) control processes to cgroupns of the MainPID= process?
1) There may no mainpid anymore
2) mainpid itself can unshare into yet another cgroupns than originally provided by systemd
(Both might be remedied if systemd kept fd to `/proc/mainpid/ns/cgroup`)

I'm mainly asking in order to simplify this predicate to mere EXEC_CGROUP and treat both cases uniformly.

Reply:
So me and @YHNdnzj discussed this a bit, and I think this the way to go eventually, but I'd like to fix this without having to implement namespace pinning in systemd.

Reply:
Good, so if you intend this PR is to be a quick fix, please add a comment here or TODO entry about possible streamlining with cgroupns FDs.